### PR TITLE
Support graceful shutdown on `serve`

### PR DESCRIPTION
**Pull request by davidpdrsn**
Currently blocked by

- https://github.com/hyperium/hyper-util/pull/66
- https://github.com/hyperium/hyper/pull/3457

But this at least proves that it works 🎉

Fixes https://github.com/tokio-rs/axum/issues/2419

### CODE REVIEW

File: axum/src/macros.rs [hunk 1]
```diff
@@ -66,3 +66,15 @@ macro_rules! all_the_tuples {
...
+macro_rules! trace {
+    ($($tt:tt)*) => {
+        tracing::trace!($($tt)*)
+    };
+}
+
+#[cfg(not(feature = "tracing"))]
+macro_rules! trace {
+    ($($tt:tt)*) => {};
+}
```
Comment (davidpdrsn) [inline axum/src/macros.rs:80]: Thinking that we should probably enabling tracing by default 🤔

File: axum/src/serve.rs [hunk 2]
```diff
@@ -308,9 +308,10 @@ where
                 tokio::spawn(async move {
                     let builder = Builder::new(TokioExecutor::new());
                     let conn = builder.serve_connection_with_upgrades(tcp_stream, hyper_service);
-                    let mut conn = pin!(conn);
+                    pin_mut!(conn);
```
Comment (jplatte) [inline axum/src/serve.rs:336]: Is this about the MSRV, or something else?

File: axum/src/serve.rs [hunk 3]
```diff
@@ -2,24 +2,29 @@
...
     net::SocketAddr,
     pin::Pin,
+    sync::Arc,
     task::{Context, Poll},
     time::Duration,
 };
 
 use axum_core::{body::Body, extract::Request, response::Response};
-use futures_util::future::poll_fn;
+use futures_util::{future::poll_fn, pin_mut, FutureExt};
```
Comment (jplatte) [inline axum/src/serve.rs:17]: While `std::pin::pin` is unfortunately not available with our MSRV, `std::future::poll_fn` is. Might make sense to use it so it becomes more apparent where we *actually* depend on the util crate.

**Comment by FauxFaux** [comment]
Don't want to be "that guy" with a :+1: comment, but, like, it may be worth pinning this issue or adding it to the release notes / upgrade guide / etc.

I looked at the graceful-shutdown example, wherein one must copy half of axum into your project, and simply decided I was going to give up on upgrading to axum 0.7.

**Comment by davidpdrsn** [comment]
> Don't want to be "that guy" with a 👍 comment, but, like, it may be worth pinning this issue or adding it to the release notes / upgrade guide / etc.

Sure. I made an issue and pinned it https://github.com/tokio-rs/axum/issues/2419

> I looked at the graceful-shutdown example, wherein one must copy half of axum into your project

Please don't make comments like that. Its quite exhausting as a maintainer to get those [all](https://github.com/tokio-rs/axum/issues/2323) [the](https://github.com/tokio-rs/axum/issues/2396) [time](https://github.com/tokio-rs/axum/discussions/2403). The changes made to hyper and hyper-util were out of our control and we're doing our best to fix it.

**Review by yanns — APPROVED** [review]
not an expert here, but it's looking good to me.
The example is so simple now 😍

**Comment by davidpdrsn** [comment]
@seanmonstar do you wanna perhaps take a look at this?

**Comment by monadbobo** [comment]
Will this MR be merged in 0.7.3?

**Comment by davidpdrsn** [comment]
@monadbobo yep!

**Comment by seanmonstar** [comment]
Looks good! I still hope to add something like https://github.com/hyperium/hyper/issues/2862. If you don't expose the inner details, you could swap to that and users wouldn't have to know :)

**Review by daheige — APPROVED** [review]
Very supportive of this design. With this smooth exit design, if you migrate version 0.6.x to version 0.7, you only need to make a few modifications to axum::Server. Your design is a very good abstract design and deserves praise.

**Comment by pashinin** [comment]
All blockers are merged. Can this be also merged?

**Comment by davidpdrsn** [comment]
Please be patient. We'll get this merged and shipped when @jplatte has approved it. There is no reason to keep commenting.

If you need it right this instant then you can depend on axum via a git dependency:

```toml
axum = { git = "https://github.com/tokio-rs/axum", branch = "david/serve-graceful-shutdown" }
```

File: examples/graceful-shutdown/Cargo.toml [hunk 4]
```diff
@@ -5,7 +5,7 @@ edition = "2021"
 publish = false
 
 [dependencies]
-axum = { path = "../../axum" }
+axum = { path = "../../axum", features = ["tracing"] }
```
[the remark this answers is not in the document]
Comment (jplatte) [reply examples/graceful-shutdown/Cargo.toml:8]: I agree.

**Review by jplatte — APPROVED** [review]
Sorry, didn't realize that this was waiting for my review 😅

Other than the missing documentation, it looks good!

**Pull request merged**